Mathlib/LinearAlgebra/Matrix/Integer.lean

```diff
@@ -0,0 +1,94 @@
+import Mathlib.Algebra.GCDMonoid.Finset
+import Mathlib.Algebra.GCDMonoid.Nat
+import Mathlib.Data.Matrix.Mul
+import Mathlib.Data.Rat.Cast.CharZero
+
+/-!
+# Lemmas on integer matrices
+
+Here we collect some results about matrices over `ℚ` and `ℤ`.
+
+## Main definitions and results
+
+* `Matrix.num`, `Matrix.den`: express a rational matrix `A` as the quotient of an integer matrix
+  by a (non-zero) natural.
+-/
+
+namespace Matrix
+
+variable {m n : Type*} [Fintype m] [Fintype n]
+
+/-!
+### Casts
+
+These results are useful shortcuts because the canonical casting maps out of `ℕ`, `ℤ`, and `ℚ` to
+suitable types are bare functions, not ring homs, so we cannot apply `Matrix.map_mul` directly to
+them.
+-/
+
+lemma map_mul_natCast {α : Type*} [NonAssocSemiring α] (A B : Matrix n n ℕ) :
+    map (A * B) ((↑) : ℕ → α) = map A (↑) * map B (↑) :=
+  Matrix.map_mul (f := Nat.castRingHom α)
+
+lemma map_mul_intCast {α : Type*} [NonAssocRing α]  (A B : Matrix n n ℤ) :
+    map (A * B) ((↑) : ℤ → α) = map A (↑) * map B (↑) :=
+  Matrix.map_mul (f := Int.castRingHom α)
+
+lemma map_mul_ratCast {α : Type*} [DivisionRing α] [CharZero α] (A B : Matrix n n ℚ) :
+    map (A * B) ((↑) : ℚ → α) = map A (↑) * map B (↑) :=
+  Matrix.map_mul (f := Rat.castHom α)
+
+/-!
+### Denominator of a rational matrix
+-/
+
+/-- The denominator of a matrix of rationals (as a `Nat`, defined as the LCM of the denominators of
+the entries). -/
+protected def den (A : Matrix m n ℚ) : ℕ := Finset.univ.lcm (fun P : m × n ↦ (A P.1 P.2).den)
+
+/-- The numerator of a matrix of rationals (a matrix of integers, defined so that
+`A.num / A.den = A`). -/
+protected def num (A : Matrix m n ℚ) : Matrix m n ℤ := ((A.den : ℚ) • A).map Rat.num
+
+lemma den_ne_zero (A : Matrix m n ℚ) : A.den ≠ 0 := by
+  simp [Matrix.den, Finset.lcm_eq_zero_iff]
+
+lemma den_dvd_iff {A : Matrix m n ℚ} {r : ℕ} :
+    A.den ∣ r ↔ ∀ i j, (A i j).den ∣ r := by
+  simp [Matrix.den]
+
+lemma num_div_den (A : Matrix m n ℚ) (i : m) (j : n) :
+    A.num i j / A.den = A i j := by
+  obtain ⟨k, hk⟩ := den_dvd_iff.mp (dvd_refl A.den) i j
+  rw [Matrix.num, map_apply, smul_apply, smul_eq_mul, mul_comm,
+    div_eq_iff <| Nat.cast_ne_zero.mpr A.den_ne_zero, hk, Nat.cast_mul, ← mul_assoc,
+    Rat.mul_den_eq_num, ← Int.cast_natCast k, ← Int.cast_mul, Rat.num_intCast]
+
+lemma inv_denom_smul_num (A : Matrix m n ℚ) :
+    (A.den⁻¹ : ℚ) • A.num.map (↑) = A := by
+  ext
+  simp [← Matrix.num_div_den A, div_eq_inv_mul]
+
+@[simp]
+lemma den_map_intCast (A : Matrix m n ℤ) : (A.map (↑)).den = 1 := by
+  simp [← Nat.dvd_one, Matrix.den_dvd_iff]
+
+@[simp]
+lemma num_map_intCast (A : Matrix m n ℤ) : (A.map (↑)).num = A := by
+  simp [Matrix.num, Function.comp_def]
+
+@[simp]
+lemma den_map_natCast (A : Matrix m n ℕ) : (A.map (↑)).den = 1 := by
+  simp [← Nat.dvd_one, Matrix.den_dvd_iff]
+
+@[simp]
+lemma num_map_natCast (A : Matrix m n ℕ) : (A.map (↑)).num = A.map (↑) := by
+  simp [Matrix.num, Function.comp_def]
+
```

Comment:
Could you also add the lemmas `den_zero`, `den_one`, `den_natCast`, `den_ofNat`, `den_intCast`, `den_ratCast`, `den_neg`, and similarly for `num`?

Reply:
I don't think `den_ratCast` applies here since `Rat.cast` only applies to division rings (which matrices are not). I've implemented the rest.

Reply:
I thought we had a direct instance somewhere, but could easily be mistaken.

Reply:
Also, it looks like you were too fast for my edit requesting `transpose`

Reply:
I dealt with `transpose`. As for `Rat.cast`, `#synth Coe ℚ  (Matrix (Fin 2) (Fin 2) ℚ)` and `#synth RatCast (Matrix (Fin 2) (Fin 2) ℚ) ` return nothing, even with all of Mathlib imported.